File: MythonJava/src/sample/Parser.java
package sample;

import sample.exceptions.IndentationException;
import sample.exceptions.InterpreterException;
import sample.exceptions.NameException;
import sample.exceptions.SyntaxException;
import sample.types.*;
import sample.types.Void;

import java.util.ArrayList;
import java.util.HashMap;

public class Parser {

    public static final int VARIABLE_NAME_INDEX = 0;
    public static final int VALUE_INDEX = 1;

    public static HashMap<String, Type> variables = new HashMap<>();

    public static final String[] PYTHON_KEYWORDS = { "and", "as", "assert", "break", "class", "continue", "def", "del",
            "elif", "else", "except", "False", "finally", "for", "from", "global", "if", "import",
            "in", "is", "lambda", "None", "nonlocal", "not", "or",
            "pass", "raise", "return", "True", "try", "while", "with", "yield" };

    /**
     * This function checks if a string is valid.
     * @param s
     * @return <code>true</code> if the string is valid, <code>false</code> otherwise.
     */
    private boolean isValidStringValue(String s) {
        if (s.charAt(0) != '\"' && s.charAt(s.length() - 1) != '\'') // If the string does not begin with quotation mark or apostrophe.
            return false;

        if (s.charAt(0) == '\"') {
            for (int i = 1; i < s.length() - 1; i++) {
                if (s.charAt(i) == '\"')
                    return false;
            }
            return s.charAt(s.length() - 1) == '\"';
        }
        for (int i = 1; i < s.length() - 1; i++) {
            if (s.charAt(i) == '\'')
                return false;
        }
        return s.charAt(s.length() - 1) == '\'';
    }

    /**
     * This function checks if a string is a list.
     * @param value
     * @return <code>true</code> if <code>value</code> is a list, false otherwise.
     */
    private boolean isList(String value) {
        return value.charAt(0) == '[' && value.charAt(value.length() - 1) == ']';
    }

    private List parseList(String value) throws InterpreterException {
        value = value.substring(1, value.length() - 1);

        if (value.length() == 0)
            return new List(new ArrayList<>());

        String[] values = value.split(",");

        for (int i = 0; i < values.length; i++)
            values[i] = values[i].strip();

        ArrayList<Type> types = new ArrayList<>();

        for (int i = 0; i < values.length; i++) {
            try {
                if (variables.containsKey(values[i])) // If the item is a variable.
                    types.add(variables.get(values[i]));
                else { // If the item is a value.
                    Type t = getType(values[i]);
                    types.add(t);
                }
            }
            catch (InterpreterException e) {
                throw e;
            }
        }
        return new List(true, types);
    }

    /**
     * This function checks the type of a value.
     * @param value A value.
     * @return An object that represents the value.
     */
    public Type getType(String value) throws InterpreterException {
        if (value == null || value.length() == 0)
            return null;

        value = value.strip();

        if (value.equals("True"))
            return new Bool(true, true);
        if (value.equals("False"))
            return new Bool(true, false);
        // If value is a boolean.

        try { // If value is an integer.
            int i = Integer.parseInt(value);
            return new Int(true, i);
        }
        catch (Exception e) { // From this point value can be a list or a string.
            if (value.equals("None")) // If the value is a null value.
                return new Str(true, null);
            if (isList(value)) { // If the value is a list.
                try {
                    return parseList(value);
                }
                catch (InterpreterException e1) { // If one of the values is not properly
                    throw e1;
                }
            }

            if (!isValidStringValue(value))
                throw new NameException(value);
            return new Str(true, value);
        }
    }

    /**
     * This function checks if a variable name is valid.
     * A valid variable name can only include digits, an underscore and english letters.
     * A valid variable name cannot start with a digit and cannot include only digits.
     * @param varName The variable name.
     * @return <code>true</code> if <code>varName</code> is valid, <code>false</code> otherwise.
     */
    public static boolean isLegalVariableName(String varName) {
        if (varName == null || varName.length() == 0)
            return false;

        if (varName.charAt(0) >= '0' && varName.charAt(0) <= '9')
            return false;

        for (String keyword : PYTHON_KEYWORDS) { // Check whether varName is a keyword.
            if (varName.equals(keyword))
                return false;
        }

        for (int i = 0; i < varName.length(); i++) {
            if (!((varName.charAt(i) >= '0' && varName.charAt(i) <= '9') ||
                    (varName.charAt(i) >= 'a' && varName.charAt(i) <= 'z') ||
                    (varName.charAt(i) >= 'A' && varName.charAt(i) <= 'Z') ||
                    varName.charAt(i) == '_'
            ))
                return false;
        }
        return true;
    }

    /**
     * This function checks if a string is a variable declaration.
     * This function assigns a value to a variable.
     * @param command A python command.
     * @return <code>true</code> if <code>command</code> is a variable declaration, false otherwise.
     */
    private boolean isAssignment(String command) throws InterpreterException {
        int firstEqualIndex = command.indexOf('=');
        int indexQuotationMarks = command.indexOf('\"');
        int indexApostrophe = command.indexOf('\'');

        if (firstEqualIndex == 0)
            throw new SyntaxException();
        if (firstEqualIndex < 0)
            return false;

        if ((firstEqualIndex < indexApostrophe || firstEqualIndex < indexQuotationMarks) || (indexApostrophe < 0 && indexQuotationMarks < 0)) { // If its a declaration on a new variable.
            String[] splitCommand = command.split("=");
            String varName = splitCommand[VARIABLE_NAME_INDEX].strip();
            String value = splitCommand[VALUE_INDEX].strip();

            try {
                if (canAdd(varName, value)) {
                    varName = varName.substring(0, varName.length() - 1).strip();
                    add(varName, value);
                    return true;
                }
            }
            catch (InterpreterException e) {
                throw e;
            }

            if (!isLegalVariableName(varName)) // If the variable name is illegal.
                throw new SyntaxException();

            if (variables.containsKey(value)) { // If it's a variable copy declaration.
                if (!variables.containsKey(value))
                    throw new NameException(varName);
                variables.put(varName, variables.get(value));
                return true;
            }
            Type t = getType(value);
            if (t == null)
                throw new SyntaxException();

            t.setTemp(false);
            variables.put(varName, t); // Add the variable.
            return true;
        }
        return false;
    }

    private boolean canAdd(String varName, String value) throws InterpreterException {
        if (!(varName.length() > 1 && varName.charAt(varName.length() - 1) == '+'))
            return false;

        varName = varName.substring(0, varName.length() - 1).strip();

        if (!variables.containsKey(varName)) // If there is a name error.
            throw new NameException(varName);
        try {
            if (variables.containsKey(value))
                return isLegalVariableName(varName);
            Type t = getType(value);
            return true;
        }
        catch (InterpreterException e) {
            throw e;
        }
    }

    /**
     * This function adds a value to a variable.
     * Note: You can only add a value to Str, Int or List variables.
     * To Str variables you can only add Str values/variables.
     * To Int variables you can only add Int values/variables.
     * To List variables you can add any values.
     * @param varName The variable to add a value to.
     * @param value The value to add.
     */
    private void add(String varName, String value) throws InterpreterException {
        Type t = variables.get(varName);

        Type update;

        if (variables.containsKey(value))
            update = variables.get(value);
        else
            update = getType(value);

        if (t instanceof List) { // If t is a list.
            ((List) t).append(getType(value));
            variables.put(varName, t);
        }

        if (t instanceof Int && update instanceof Int) {
            ((Int) t).setValue(((Int) t).getValue() + ((Int) update).getValue());
            variables.put(varName, t);
        }

        if (t instanceof Str && update instanceof Str) {
            String current = ((Str) t).getValue();
            String s = ((Str) update).getValue();
            ((Str) t).setValue(current.substring(0, current.length() - 1) + s.substring(1));
            variables.put(varName, t);
        }
    }

    /**
     * This function checks if a command is a variable print declaration.
     * @param command
     * @return
     */
    public boolean isVariablePrint(String command) throws InterpreterException {
        if (!variables.containsKey(command))
            throw new NameException(command);
        return variables.get(command).isPrintable();
    }

    /**
     * This function parses a python command.
     * @param command A python command.
     */
    public Type parseString(String command) throws InterpreterException {
        if (command.equals(""))
            return null;

        if (command.charAt(0) == ' ' || command.charAt(0) == '\t')
            throw new IndentationException();
        command = command.strip();

        try {
            boolean isAssignment = isAssignment(command);
            boolean isDeleteStatement = isDeleteStatement(command);
            boolean isTypeStatement = isTypeStatement(command);
            boolean isLenStatement = isLenStatement(command);

            if (isLenStatement) {
                len(command);
                return null;
            }

            if (isDeleteStatement) {
                delete(command);
                return null;
            }

            if (isTypeStatement) {
                printType(command);
                return null;
            }

            if (isAssignment)
                return new Void();
            else {
                boolean isVariablePrint = isVariablePrint(command);
                if (isVariablePrint)
                    System.out.println(variables.get(command).toString());
            }
        }
        catch (InterpreterException e) {
            throw e;
        }
        return null;
    }

    /**
     * This function checks if a command is a delete command.
     * @param command A python command.
     * @return <code>true</code> if the command is a delete command, <code>false</code> otherwise.
     */
    private boolean isDeleteStatement(String command) {
        return command.startsWith("del ");
    }

    /**
     * This function deletes a variable from <code>variables</code>.
     * @param command A python command.
     */
    private void delete(String command) {
        String value = command.substring(4).strip();

        if (variables.containsKey(value))
            variables.remove(value);
        else
            new NameException(value).printStackTrace();
    }

    /**
     * This function checks if a statement is a <code>type(value)</code> statement.
     * @param command A python command.
     * @return <code>true</code> if the command is a <code>type(value)</code>, false otherwise.
     */
    private boolean isTypeStatement(String command) {
        return command.startsWith("type(") && command.endsWith(")");
    }

    private void printType(String command) {
        String value = command.substring(5, command.length() - 1).strip(); // Get the value inside the function.

        if (value.equals("")) {
            new SyntaxException().printStackTrace();
            return;
        }

        if (variables.containsKey(value)) { // If the user wants to check the type of a variable.
            System.out.println(variables.get(value).getType());
            return;
        }
        try { // If the user wants to check the type of a value.
            Type t = getType(value);
            System.out.println(t.getType());
        }
        catch (InterpreterException e) {
            e.printStackTrace();
        }
    }

    private boolean isLenStatement(String command) {
        return command.startsWith("len(") && command.endsWith(")");
    }

    private void len(String command) throws InterpreterException{
        String value = command.substring(4, command.length() - 1).strip();

        if (value.equals(""))
            throw new SyntaxException();

        Type t;

        if (variables.containsKey(value)) // If the user wants to check the length of a variable.
            t = variables.get(value);
        else { // If the user wants to check the length of a value.
            try {
                t = getType(value);
            } catch (InterpreterException e) {
                e.printStackTrace();
                return;
            }
        }
        if (t instanceof Str)
            System.out.println(((Str)t).len());
        else if (t instanceof List)
            System.out.println(((List)t).len());
        else
            throw new SyntaxException();
    }
}

File: MythonC++/Mython/Parser.cpp
#include <iostream>
#include <string>
#include "math.h"
#include "Parser.h"
#include "IndentationException.h"
#include "SyntaxException.h"
#include "Type.h"
#include "Boolean.h"
#include "Integer.h"
#include "String.h"
#include "Void.h"

using namespace std;

/// <summary>
/// This function counts how many times c appears in s.
/// </summary>
/// <param name="s"></param>
/// <param name="c"></param>
/// <returns></returns>
int Parser::Count(string s, char c)
{
    int count = 0;
    for (int i = 0; i < s.length(); i++)
    {
        if (s[i] == c)
            count++;
    }
    return count;
}

vector<string> Parser::ParseListString(string value)
{
    int index = 0;
    vector<string> items = vector<string>();
    for (int i = 0; i < value.length(); i++)
    {
        string s = "";
        while (i < value.length() && value[i] != ',')
        {
            s += value[i];
            i++;
        }
            items.push_back(RemoveSpaces(s));
    }
    return items;
}

/// <summary>
/// This function splits a string based on character delimiter.
/// </summary>
/// <param name="command">A command to split</param>
/// <param name="delimiter">A char delimiter.</param>
/// <returns>Split array.</returns>
/// <example>
///     >>> SplitStrings("x = 78", ' ')
///     >>> {"x", "=", "78"}
/// </example>
void Parser::SplitString(string split[], string command, char delimiter)
{
    string value = "";
    int index = 0;

    if (command.find_first_of('=') != command.find_last_of('='))
        return;

    while (index < command.length() && command[index] != delimiter) // Get the first slice.
    {
        value += command[index];
        index++;
    }
    while (index < command.length() && command[index] == delimiter) { index++; } // Skip all the delimiter characters.

    split[0] = value;
    value = "";

    while (index < command.length() && command[index] != delimiter) // Get the second slice.
    {
        value += command[index];
        index++;
    }
    split[1] = value;
}

/// <summary>
/// This function removes all the spaces at the beginning and the end of a string.
/// </summary>
/// <param name="value"></param>
/// <returns></returns>
string Parser::RemoveSpaces(string value)
{
    int beginIndex = 0;
    int stopIndex = value.length() - 1;

    while (beginIndex < value.length() && value[beginIndex] == ' ')
        beginIndex++;
    while (stopIndex >= 0 && value[stopIndex] == ' ')
        stopIndex--;

    return value.substr(beginIndex, stopIndex + 1);
}

bool Parser::IsList(string value)
{
    return value[0] == '[' && value[value.length() - 1] == ']';
}

/// <summary>
/// This function parses a list.
/// </summary>
/// <param name="value"></param>
/// <returns></returns>
Type* Parser::ParseList(string value)
{
    List* lst;

    if (value == "[]")
        return new List(true, vector<Type*>());

    value = value.substr(1, value.length() - 2);
        
    vector<string> items = ParseListString(value);
    vector<Type*> list;

    for (int i = 0; i < items.size(); i++)
    {
        try
        {
            Type* t = GetType(items[i]);
            list.push_back(t);
        }
        catch (InterperterException& e)
        {
            throw e;
        }
    }
    lst = new List(true, list);
    return lst;
}

/// <summary>
/// This function checks if a command is a variable declaration.
/// </summary>
/// <param name="command"></param>
/// <returns></returns>
bool Parser::MakeAssignment(string command)
{
    int commandLength = command.length();

    int indexEqual = command.find('=');
    int indexQuotationMarks = command.find('\"');
    int indexApostrophe = command.find('\'');

    string split[] = { "", "" };

    if (indexEqual < 0)
        return false;

    if (indexEqual > 0) // Check if a command is a vairable copy command.
    {
        SplitString(split, command, '=');

        string varName = RemoveSpaces(split[VAR_NAME_INDEX]);
        string value = RemoveSpaces(split[VALUE_INDEX]);

        if (varName == "" && value == "")
            throw SyntaxException();

        if (Parser::CanAdd(varName, value)) // If it's an addition statement.
        {
            try
            {
                Add(varName, value);
                return true;
            }
            catch (InterperterException& e)
            {
                throw e;
            }
        }

        if (!IsLegalVarName(varName))
            throw NameException(varName);

        if (IsVariableName(value))
        {
            if (dynamic_cast<List*>(variables[varName]) != NULL) // If the value is not a list we want to create a clone of the variable.
                variables[varName] = variables[value]->Clone();
            else // If the variable is a list we want a direct pointer to the original list.
                variables[varName] = variables[value];

            return true;
        }
        else
        {
            try
            {
                Type* t = GetType(value);
                t->SetTemp(false);
                variables[varName] = t;
                return true;
            }
            catch (SyntaxException& e)
            {
                throw e;
            }
        }
    }
    return false;
}

/// <summary>
/// This function parses a command in Python.
/// </summary>
/// <param name="command">A command in Python.</param>
/// <returns>The output of the command.</returns>
Type* Parser::ParseString(string command)
{
    Type* t;

    if (command.length() == 0)
        return NULL;

    if (command[0] == ' ' || command[0] == '\t') // If the line begins wil an illegal indentation.
        throw IndentationException();

    command = RemoveSpaces(command);

    if (IsDeleteStatement(command)) // If the command is a delete command.
    {
        try
        {
            Delete(command);
            return NULL;
        }
        catch (InterperterException &e)
        {
            throw e;
        }
    }

    if (IsTypeStatement(command)) // If the command is "type(value)".
    {
        PrintType(command);
        return NULL;
    }

    if (IsLenStatement(command))
    {
        try
        {
            Len(command);
            return NULL;
        }
        catch (InterperterException& e)
        {
            cout << e.what() << endl;
            return NULL;
        } 
    }

    bool isAssignment =  MakeAssignment(command);
    bool isValuePrint = IsValuePrint(command);

    if (isValuePrint)
    {
        Type* t = GetType(command);
        cout << t->ToString() << endl;
        return NULL;
    }

    if (isAssignment) // If the given command is an assignment
    {
        t = new Void(true);
        return t;
    }
    else // If it's a print variable statemnt.
    {
        Type* t = GetVariableValue(command);

        if (t == nullptr)
            throw NameException(command);
        else
            cout << t->ToString() << endl;
    }
    return NULL;
}

/// <summary>
/// This function returns the type of a value.
/// </summary>
/// <param name="value">A value. Can be int, string or boolean</param>
/// <returns>The class that represents this value.</returns>
Type* Parser::GetType(string value)
{
    Type* type;

    if (value.length() == 0) // If the value is empty we will return a null value.
        throw SyntaxException();

    // Check if the values are boolean values.
    if (value == "True")
    {
        type = new Boolean(true, true);
        return type;
    }
    else if (value == "False")
    {
        type = new Boolean(false, true);
        return type;
    }
    // From here the value can be a string or an int.

    int numericValue;

    try // Try to cast the value to an integer.
    {
        numericValue = stoi(value);
    }
    catch (exception& e) // If the cast fails then the value must be a string or a list.
    { 
        if (IsList(value)) // If the value is a list.
        {
            try { // Parse the list.
                type = ParseList(value);
                return type;
            }
            catch (InterperterException& e) // There is an error with the list.
            {
                throw e;
            }
        }

        if (!String::IsValidString(value))
            throw SyntaxException();

        type = new String(value, true);
        return type;
    }
    type = new Integer(numericValue, true);
    return type;
}

/// <summary>
/// This function checks if a variable name is valid.
/// A valid variable name can only include english letters, digits and an underscore.
/// The first character must be an underscore or an english letter. A valid variable name cannot include digits only.
/// In addition, a valid variable name cannot be a python keyword.
/// </summary>
/// <param name="varName">The name of the variable</param>
/// <returns>true if varName is valid, false otherwise.</returns>
bool Parser::IsLegalVarName(string varName)
{
    for (string s : pythonKeywords)
    {
        if (varName == s) // If varName is a python keyword then it's not a valid variable name.
            return false;
    }

    int countDigit = 0;

    for (char c : varName)
    {
        if (!((c >= '0' && c <= '9') || (c == '_') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))) // If one of the letters in varName is not valid.
            return false;
        if (c >= '0' && c <= '9')
            countDigit++;
    }
    return countDigit != varName.length() && !(varName[0] >= '0' && varName[0] <= '9'); // Check if the first character is an underscore or a letter, and if not all the characters in varName are digits.
}

Type* Parser::GetVariableValue(string varName)
{
    return variables[varName];
}
/// <summary>
/// This function clears variables.
/// </summary>
void Parser::RemoveVariables()
{
    for (pair<string, Type*> p : variables)
        delete p.second;
}

/// <summary>
/// This function checks whether a string is a variable name.
/// </summary>
/// <param name="varName"></param>
/// <returns></returns>
bool Parser::IsVariableName(string varName)
{
    for (pair<string, Type*> p : variables)
    {
        if (p.first == varName)
            return true;
    }
    return false;
}

/// <summary>
/// This function checks is a declatation is an addition statement.
/// </summary>
/// <param name="varName"></param>
/// <param name="value"></param>
/// <returns></returns>
bool Parser::CanAdd(string varName, string value)
{
    if (varName[varName.length() - 1] != '+')
        return false;
    return IsVariableName(value) || GetType(value) != NULL;
}

/// <summary>
/// This function adds a value to a list or to an integer or a string.
/// </summary>
/// <param name="varName"></param>
/// <param name="value"></param>
/// <returns></returns>
void Parser::Add(string varName, string value)
{
    varName = RemoveSpaces(varName.substr(0, varName.length() - 1));
    if (!IsVariableName(varName)) // If the isn't such variable name.
        throw NameException(varName);
    Type* t;

    if (!IsVariableName(value)) // If its an addition of a value.
    {
        try
        {
            t = GetType(value);
        }
        catch (InterperterException &e)
        {
            throw e;
        }
    }
    else // If its an addition of an existing values.
    {
        t = variables[value];
    }

    String* varNameStringValue = dynamic_cast<String*>(variables[varName]);
    Integer* varNameIntegerValue = dynamic_cast<Integer*>(variables[varName]);
    List* varNameListValue = dynamic_cast<List*>(variables[varName]);

    String* valueString = dynamic_cast<String*>(t);
    Integer* valueInteger = dynamic_cast<Integer*>(t);

    if (varNameStringValue != NULL && valueString != NULL) // If the value is a string
    {
        string varNameStr = varNameStringValue->GetValue();
        varNameStringValue->SetValue(varNameStr.substr(0, varNameStr.length() - 1) + valueString->GetValue().substr(1));
        variables[varName] = varNameStringValue;
        return;
    }
    else if (varNameIntegerValue != NULL && valueInteger != NULL) // If the value is an integer.
    {
        varNameIntegerValue->SetValue(varNameIntegerValue->GetValue() + valueInteger->GetValue());
        variables[varName] = varNameIntegerValue;
        return;
    }
    else if (varNameListValue != NULL) // If we wna tot append an item to the list.
    {
        varNameListValue->Append(t);
        return;
    }
// TODO - add list.
    throw SyntaxException();
}

/// <summary>
/// This function checks if the given command is a print value statement.
/// </summary>
/// <param name="command"></param>
/// <returns></returns>
bool Parser::IsValuePrint(string command)
{
    try
    {
        Type* t = GetType(command);
        return true;
    }
    catch (const std::exception&)
    {
        return false;
    }
}

/// <summary>
/// This function checks if a command is a command from this form: type(value).
/// </summary>
/// <param name="command"></param>
/// <returns></returns>
bool Parser::IsTypeStatement(string command)
{
    if (command.length() < 6)
        return false;
    string type = command.substr(0, 5);
    return type == "type(" && command[command.length() - 1] == ')'; // Check if the command begin with "type(" and ends with a ")".
}

/// <summary>
/// This function prints the type of a value.
/// </summary>
/// <param name="command"></param>
void Parser::PrintType(string command)
{
    if (command == "type()")
        return;

    string value = command.substr(5, command.length() - 1);
    value = value.substr(0, value.length() - 1);
    value = RemoveSpaces(value);


    if (IsVariableName(value)) // If the user wants to check the type of a variable.
    {

        cout << variables[value]->GetType() << endl;
    }

    else
    {
        try
        {
            Type* t = GetType(value);
            cout << t->GetType() << endl;
        }
        catch (InterperterException &e)
        {
            cout << e.what() << endl;
        }
    }
}
/// <summary>
/// This function checks if a statement is a delete statement.
/// </summary>
/// <param name="command"></param>
/// <returns></returns>
bool Parser::IsDeleteStatement(string command)
{
    if (command.length() < 5) // the command must be at least five characters long("del varName").
        return false;

    string del = command.substr(0, 3);
    return del == "del" && command[3] == ' '; // After the keyword a space must be present.
}

void Parser::Delete(string command)
{
    string varName = command.substr(4);
    varName = RemoveSpaces(varName);

    if (!IsVariableName(varName))
        throw NameException(varName);
    
    variables.erase(varName);
}

/// <summary>
/// This function checks if a statement is a length statement.
/// </summary>
/// <param name="command"></param>
/// <returns></returns>
bool Parser::IsLenStatement(string command)
{
    if (command.length() < 4)
        return false;
    return command.substr(0, 4) == "len(";
}

/// <summary>
/// This function finds the length of a given variables.
/// If the variable is not a string or a list, the function will throw an exception.
/// </summary>
/// <param name="command"></param>
void Parser::Len(string command)
{
    string value = command.substr(4);
    value = value.substr(0, value.length() - 1);
    value = RemoveSpaces(value);
    Type* t;

    if (IsVariableName(value)) // If the argument is an existing variable.
        t = variables[value];

    else // If the argument is a value.
    {
        try
        {
            t = GetType(value);
        }
        catch (InterperterException &e)
        {
            cout << e.what();
            return;
        }
    }
    
    if (t->GetType() == TYPE_LIST) // If the value is a list.
    {
        List* list = dynamic_cast<List*>(t);
        cout << list->Len() << endl;
    }

    else if (t->GetType() == TYPE_STR) // If the value is a string.
    {
        String* str = dynamic_cast<String*>(t);
        cout << str->Len() << endl;
    }
    else // The value isn't string nor a list, so the function will throw an exception.
        throw SyntaxException();
}
